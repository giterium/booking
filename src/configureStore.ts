import { createBrowserHistory } from 'history'
import thunk from 'redux-thunk';
import { applyMiddleware, compose, createStore } from 'redux'
import createRootReducer from './reducers'

export const history = createBrowserHistory()

export default function configureStore(preloadedState) {
    const composeEnhancer =  (window && (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) || compose;
    const store = createStore(
        createRootReducer(history),
        preloadedState,
        composeEnhancer(
            applyMiddleware(
                thunk
            ),
        ),
    )
    return store
}

export const store = configureStore([]);