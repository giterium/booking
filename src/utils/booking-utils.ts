import moment, { Moment } from "moment"
import styles from "../css/booking.module.css"
import { store } from '../configureStore'
import {TypeBooking} from "../reducers/booking-reducers"
import { createAction } from 'redux-actions'
import {TypeSelected} from "../reducers/selected-reducers"
import {isGoodRange} from "./isGoodRange"

export const updateSelected = createAction(
    'UPDATED_SELECTED',
    updates => updates,
);

export const setCurrentBooking = createAction(
    'UPDATED_CURRENT_BOOKING',
    updates => updates,
);

export function momentNullDate ():Moment {
    return moment("1970-01-01 00:00")
}

const unSelected = () => {
    store.dispatch(updateSelected({
        start:{day:momentNullDate(), room:''},
        end:{day:momentNullDate(), room:''}
    }))
}

export function timenull(date: Date | Moment):(Date | Moment) {
    if(moment.isMoment(date))
        return date.clone().milliseconds(0).second(0).minutes(0).hours(0)
    else
        return moment(date).clone().milliseconds(0).second(0).minutes(0).hours(0).toDate()
}

export const isBooking = (day, room) => {
    const currentBooking:TypeBooking = store.getState().currentBooking
    const booking:TypeBooking[] = store.getState().booking

    const bookingWithCurrent = [...booking]
    const index = booking.findIndex((p) => p._id == currentBooking._id)
    if(index != -1) {
        bookingWithCurrent[index] = {...currentBooking}
    }
    const bookingSlice = bookingWithCurrent.filter(item => item.room == room._id)

    let is_booking = false
    let is_first_booking = false
    let is_last_booking = false

    for(const curBooking of bookingSlice) {
        if (
            moment(timenull(curBooking.startDate)).diff(timenull(day.date)) <= 0
            &&
            moment(timenull(curBooking.endDate)).diff(timenull(day.date)) >= 0
            &&
            curBooking.room == room._id
        ) {
            if(moment(timenull(curBooking.startDate)).diff(timenull(day.date)) == 0) {
                is_first_booking = true
            }
            if(moment(timenull(curBooking.endDate)).diff(timenull(day.date)) == 0) {
                is_last_booking = true
            }
            is_booking = true
        }
    }
    if(is_booking) {
        if(is_last_booking && is_first_booking)
            return styles.isBooking + ' '+styles.isFirstBooking+ ' '+styles.isLastBooking
        else if(is_first_booking)
            return styles.isBooking + ' '+styles.isFirstBooking
        else if(is_last_booking)
            return styles.isBooking + ' '+styles.isLastBooking
        else
            return styles.isBooking
    }
    else
        return ''
}

export const commonStylesCell = (day, room) => {
    return isBooking(day, room) + ' cell d' + day.date.format('DD-MM-YY') + ' ' +styles.dayCell
}

export const isSelected = (day, room, selected) => {
    if(selected.start.room && !selected.end.room ) {
        if(selected.start.day.format('DD.MM.YY') == day.date.format('DD.MM.YY') && selected.start.room == room._id) {
            return styles.selected +' '+ styles.firstSelected
        }
        else {
            return ''
        }
    }
    else if(selected.start.room == room._id && selected.end.room == room._id) {
        if(selected.start.day.diff(day.date) <= 0 && selected.end.day.diff(day.date) >= 0) {
            if(selected.start.day.format('DD.MM.YY') == day.date.format('DD.MM.YY')) {
                return styles.selected + ' ' + styles.firstSelected
            }
            else if(moment(timenull(selected.end.day)).diff(timenull(day.date)) == 0) {
                return styles.selected + ' ' + styles.lastSelected
            }
            else
                return styles.selected
        }
        else {
            return ''
        }
    }
    else {
        return ''
    }
}

export const clickDayRoom = (day, room) => {
    const dispatch = store.dispatch;
    const selected:TypeSelected = store.getState().selected;

    if(isBooking(day, room) != styles.isBooking && isBooking(day, room) != styles.isBooking + ' '+styles.isFirstBooking+ ' '+styles.isLastBooking) {
        if (!selected.start.room) {
            dispatch(updateSelected({
                start: {day: day.date, room: room._id},
                end: {day: momentNullDate(), room: ''}
            }))
        } else {
            if (!selected.end.room) {
                if(isGoodRange(selected.start.day, day.date, room._id)) {
                    if (selected.start.day.format('DD.MM.YY') == day.date.format('DD.MM.YY')) {
                        unSelected();
                    } else if (selected.start.room != room._id)
                        unSelected();
                    else {
                        if (day.date < selected.start.day) {
                            dispatch(updateSelected({
                                start: {day: day.date, room: room._id},
                                end: selected.start
                            }))
                        }
                        else {
                            dispatch(updateSelected({
                                start: selected.start,
                                end: {day: day.date, room: room._id}

                            }))
                        }
                    }
                }
                else {
                    unSelected()
                }
            }
            else {
                unSelected()
            }
        }
    }
    else {
        unSelected()
    }
};