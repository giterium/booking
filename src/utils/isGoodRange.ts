import {TypeBooking} from "../reducers/booking-reducers";
import {store} from "../configureStore";
import moment, { Moment} from "moment";
import {timenull} from "./booking-utils";

export function isGoodRange (startDate: Date | Moment, endDate: Date | Moment, id_room: string, curIdBooking: string = ''): boolean {
    const booking:TypeBooking[] = store.getState().booking;
    const bookingSlice = booking.filter(item => item.room == id_room)
    for(const curBooking of bookingSlice) {
        if(
            (
                moment(timenull(startDate)).isAfter(timenull(curBooking.startDate)) && moment(timenull(startDate)).isBefore(timenull(curBooking.endDate))
                ||
                moment(timenull(endDate)).isAfter(timenull(curBooking.startDate)) && moment(timenull(endDate)).isBefore(timenull(curBooking.endDate))
                ||
                moment(timenull(startDate)).isSameOrBefore(timenull(curBooking.startDate)) && moment(timenull(endDate)).isSameOrAfter(timenull(curBooking.endDate))
                ||
                moment(timenull(startDate)).isSame(timenull(curBooking.startDate)) && moment(timenull(endDate)).isSame(timenull(curBooking.endDate))
                ||
                moment(timenull(startDate)).isSameOrAfter(timenull(curBooking.endDate)) && moment(timenull(endDate)).isSameOrBefore(timenull(curBooking.startDate))
            )
            &&
            curIdBooking != curBooking._id
        ) {
            return false;
        }
    }
    return true;
}