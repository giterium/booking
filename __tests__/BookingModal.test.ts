import React, { useContext } from "react";
import { BookingModal } from "../src/views/BookingModal";
import { Provider } from "react-redux";
import configureStore from 'redux-mock-store';
const mockStore = configureStore([]);
import "regenerator-runtime/runtime";
import {momentNullDate} from "../src/utils/booking-utils";
import {Booking} from "../src/views/Booking";

describe("Booking Modal component", () => {
    let store;
    const setState = jest.fn();
    const useStateSpy = jest.spyOn(React, "useState")
    useStateSpy.mockImplementation((init) => [init, setState]);

    beforeEach(() => {
        store = mockStore({
            rooms: [{
                _id: 'enzymeOpenWindow',
                num: 22,
                building: 2,
                capacity: 2,
                cost: '100'
            },{
                _id: '123',
                num: 33,
                building: 2,
                capacity: 2,
                cost: '200'
            }],
            selected: {
                start:{day:momentNullDate(), room:''},
                end:{day:momentNullDate(), room:''}
            },
            currentBooking: {}
        });

    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    it("should snapshot BookingModal ", () => {
        const wrapper = mount(<Provider store={store}><BookingModal /></Provider>);
        expect(wrapper.debug()).toMatchSnapshot();
    });

    it('renders modal when open flag is true', () => {
        const wrapper = mount(
            <Provider store={store}><BookingModal /></Provider>
        );
        expect(wrapper.find('.titleModal').exists()).toEqual(true);
    });

    it("should call onActionClose method", () => {
        const mockCallBack = jest.fn();
        const wrapper = mount( <Provider store={store}><BookingModal onActionClose={mockCallBack} /></Provider>)
        expect(mockCallBack.mock.calls.length).toBe(0);
        wrapper.find(".closeButton").last().simulate("click");
        expect(mockCallBack.mock.calls.length).toBe(1);
    });

    it("should call onActionClose method", () => {
        const mockCallBack = jest.fn();
        const wrapper = mount( <Provider store={store}><BookingModal onActionModal={mockCallBack} /></Provider>)
        expect(mockCallBack.mock.calls.length).toBe(0);
        wrapper.find(".actionButton").last().simulate("click");
        expect(mockCallBack.mock.calls.length).toBe(1);
    });

    it('datepicker calls the onChange Event in StartDate', done => {
        const mockCallBack = jest.fn();
        const wrapper = mount( <Provider store={store}><BookingModal onChangeBooking={mockCallBack} /></Provider>)
        const dateInput = wrapper.find(".startDatePicker").last();
        dateInput.simulate('change', {target: { value: "2018-01-04" }});
        done();
        expect(mockCallBack.mock.calls.length).toBe(1);
    });

    it('datepicker calls the onChange Event in EndDate', done => {
        const mockCallBack = jest.fn();
        const wrapper = mount( <Provider store={store}><BookingModal onChangeBooking={mockCallBack} /></Provider>)
        const dateInput = wrapper.find(".endDatePicker").last();
        dateInput.simulate('change', {target: { value: "2018-01-04" }});
        done();
        expect(mockCallBack.mock.calls.length).toBe(1);
    });

    it("should call onActionDelete method", () => {
        const mockCallBack = jest.fn();
        const wrapper = mount( <Provider store={store}><BookingModal onActionDelete={mockCallBack} /></Provider>)
        expect(mockCallBack.mock.calls.length).toBe(0);
        wrapper.find(".deleteButton").last().simulate("click");
        expect(mockCallBack.mock.calls.length).toBe(1);
    });

    it("should render Cost default value", () => {
        const wrapper = mount( <Provider store={store}><BookingModal /></Provider>)
        const errorBox = wrapper.find(".cost");
        expect(errorBox.text()).toBe('0');
    });

});