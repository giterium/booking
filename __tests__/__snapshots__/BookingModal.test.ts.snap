// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Booking Modal component should snapshot BookingModal  1`] = `
"<Provider store={{...}}>
  <BookingModal>
    <div id=\\"root\\">
      <Modal isOpen={true} style={{...}} ariaHideApp={false} portalClassName=\\"ReactModalPortal\\" bodyOpenClassName=\\"ReactModal__Body--open\\" role=\\"dialog\\" closeTimeoutMS={0} shouldFocusAfterRender={true} shouldCloseOnEsc={true} shouldCloseOnOverlayClick={true} shouldReturnFocusAfterClose={true} preventScroll={false} parentSelector={[Function: parentSelector]} overlayElement={[Function: overlayElement]} contentElement={[Function: contentElement]}>
        <Portal containerInfo={{...}}>
          <ModalPortal defaultStyles={{...}} isOpen={true} style={{...}} ariaHideApp={false} portalClassName=\\"ReactModalPortal\\" bodyOpenClassName=\\"ReactModal__Body--open\\" role=\\"dialog\\" closeTimeoutMS={0} shouldFocusAfterRender={true} shouldCloseOnEsc={true} shouldCloseOnOverlayClick={true} shouldReturnFocusAfterClose={true} preventScroll={false} parentSelector={[Function: parentSelector]} overlayElement={[Function: overlayElement]} contentElement={[Function: contentElement]}>
            <div className=\\"ReactModal__Overlay ReactModal__Overlay--after-open\\" style={{...}} onClick={[Function (anonymous)]} onMouseDown={[Function (anonymous)]}>
              <div id={[undefined]} style={{...}} className=\\"ReactModal__Content ReactModal__Content--after-open\\" tabIndex=\\"-1\\" onKeyDown={[Function (anonymous)]} onMouseDown={[Function (anonymous)]} onMouseUp={[Function (anonymous)]} onClick={[Function (anonymous)]} role=\\"dialog\\" aria-label={[undefined]} aria-modal={true} data-testid={[undefined]}>
                <h2 className=\\"titleModal\\">
                   
                </h2>
                <br />
                <table className=\\"tableModal\\">
                  <tbody>
                    <TableInput title=\\"Name\\" className=\\"usualInput\\" name=\\"fio\\" changeUpdate={[Function: changeUpdate]} defaultValue=\\"\\">
                      <tr>
                        <td className=\\"updateFieldTitle\\">
                          Name
                          :
                        </td>
                        <td>
                          <input type=\\"text\\" className=\\"usualInput\\" name=\\"fio\\" onChange={[Function: changeUpdate]} onFocus={[undefined]} onClick={[undefined]} onKeyUp={[Function: changeUpdate]} onKeyPress={[undefined]} defaultValue=\\"\\" />
                        </td>
                      </tr>
                    </TableInput>
                    <TableErrors errors=\\"\\">
                      <tr>
                        <td colSpan={2}>
                          <div className=\\"errorBox\\" />
                        </td>
                      </tr>
                    </TableErrors>
                    <tr>
                      <td className=\\"cellModal\\">
                        Date of entry: 
                      </td>
                      <td>
                        <r className=\\"startDatePicker\\" locale=\\"ru\\" selected={{...}} onChange={[Function: onChange]} dateFormat=\\"dd.MM.yyyy\\" allowSameDay={false} dateFormatCalendar=\\"LLLL yyyy\\" disabled={false} disabledKeyboardNavigation={false} dropdownMode=\\"scroll\\" onFocus={[Function: onFocus]} onBlur={[Function: onBlur]} onKeyDown={[Function: onKeyDown]} onInputClick={[Function: onInputClick]} onSelect={[Function: onSelect]} onClickOutside={[Function: onClickOutside]} onMonthChange={[Function: onMonthChange]} onCalendarOpen={[Function: onCalendarOpen]} onCalendarClose={[Function: onCalendarClose]} preventOpenOnFocus={false} onYearChange={[Function: onYearChange]} onInputError={[Function: onInputError]} monthsShown={1} readOnly={false} withPortal={false} shouldCloseOnSelect={true} showTimeSelect={false} showTimeInput={false} showPreviousMonths={false} showMonthYearPicker={false} showFullMonthYearPicker={false} showTwoColumnMonthYearPicker={false} showYearPicker={false} showQuarterYearPicker={false} strictParsing={false} timeIntervals={30} timeCaption=\\"Time\\" previousMonthButtonLabel=\\"Previous Month\\" nextMonthButtonLabel=\\"Next Month\\" previousYearButtonLabel=\\"Previous Year\\" nextYearButtonLabel=\\"Next Year\\" timeInputLabel=\\"Time\\" enableTabLoop={true} yearItemNumber={12} renderDayContents={[Function: renderDayContents]} focusSelectedMonth={false} showPopperArrow={true} excludeScrollbar={true} customTimeInput={{...}}>
                          <r className={[undefined]} wrapperClassName={[undefined]} hidePopper={true} portalId={[undefined]} popperModifiers={{...}} targetComponent={{...}} popperContainer={[undefined]} popperComponent={{...}} popperPlacement=\\"bottom-start\\" popperProps={{...}} popperOnKeyDown={[Function (anonymous)]} enableTabLoop={true}>
                            <Manager className=\\"react-datepicker-manager\\">
                              <Reference>
                                <InnerReference setReferenceNode={[Function (anonymous)]}>
                                  <div className=\\"react-datepicker-wrapper\\">
                                    <div className=\\"react-datepicker__input-container\\">
                                      <input type=\\"text\\" value=\\"06.02.2021\\" onBlur={[Function (anonymous)]} onChange={[Function (anonymous)]} onClick={[Function (anonymous)]} onFocus={[Function (anonymous)]} onKeyDown={[Function (anonymous)]} id={[undefined]} name={[undefined]} autoFocus={[undefined]} placeholder={[undefined]} disabled={false} autoComplete={[undefined]} className=\\"startDatePicker\\" title={[undefined]} readOnly={false} required={[undefined]} tabIndex={[undefined]} aria-describedby={[undefined]} aria-invalid={[undefined]} aria-labelledby={[undefined]} aria-required={[undefined]} />
                                    </div>
                                  </div>
                                </InnerReference>
                              </Reference>
                            </Manager>
                          </r>
                        </r>
                      </td>
                    </tr>
                    <TableErrors errors=\\"\\">
                      <tr>
                        <td colSpan={2}>
                          <div className=\\"errorBox\\" />
                        </td>
                      </tr>
                    </TableErrors>
                    <tr>
                      <td className=\\"cellModal\\">
                        Date of departure: 
                      </td>
                      <td>
                        <r className=\\"endDatePicker\\" locale=\\"ru\\" selected={{...}} onChange={[Function: onChange]} dateFormat=\\"dd.MM.yyyy\\" allowSameDay={false} dateFormatCalendar=\\"LLLL yyyy\\" disabled={false} disabledKeyboardNavigation={false} dropdownMode=\\"scroll\\" onFocus={[Function: onFocus]} onBlur={[Function: onBlur]} onKeyDown={[Function: onKeyDown]} onInputClick={[Function: onInputClick]} onSelect={[Function: onSelect]} onClickOutside={[Function: onClickOutside]} onMonthChange={[Function: onMonthChange]} onCalendarOpen={[Function: onCalendarOpen]} onCalendarClose={[Function: onCalendarClose]} preventOpenOnFocus={false} onYearChange={[Function: onYearChange]} onInputError={[Function: onInputError]} monthsShown={1} readOnly={false} withPortal={false} shouldCloseOnSelect={true} showTimeSelect={false} showTimeInput={false} showPreviousMonths={false} showMonthYearPicker={false} showFullMonthYearPicker={false} showTwoColumnMonthYearPicker={false} showYearPicker={false} showQuarterYearPicker={false} strictParsing={false} timeIntervals={30} timeCaption=\\"Time\\" previousMonthButtonLabel=\\"Previous Month\\" nextMonthButtonLabel=\\"Next Month\\" previousYearButtonLabel=\\"Previous Year\\" nextYearButtonLabel=\\"Next Year\\" timeInputLabel=\\"Time\\" enableTabLoop={true} yearItemNumber={12} renderDayContents={[Function: renderDayContents]} focusSelectedMonth={false} showPopperArrow={true} excludeScrollbar={true} customTimeInput={{...}}>
                          <r className={[undefined]} wrapperClassName={[undefined]} hidePopper={true} portalId={[undefined]} popperModifiers={{...}} targetComponent={{...}} popperContainer={[undefined]} popperComponent={{...}} popperPlacement=\\"bottom-start\\" popperProps={{...}} popperOnKeyDown={[Function (anonymous)]} enableTabLoop={true}>
                            <Manager className=\\"react-datepicker-manager\\">
                              <Reference>
                                <InnerReference setReferenceNode={[Function (anonymous)]}>
                                  <div className=\\"react-datepicker-wrapper\\">
                                    <div className=\\"react-datepicker__input-container\\">
                                      <input type=\\"text\\" value=\\"06.02.2021\\" onBlur={[Function (anonymous)]} onChange={[Function (anonymous)]} onClick={[Function (anonymous)]} onFocus={[Function (anonymous)]} onKeyDown={[Function (anonymous)]} id={[undefined]} name={[undefined]} autoFocus={[undefined]} placeholder={[undefined]} disabled={false} autoComplete={[undefined]} className=\\"endDatePicker\\" title={[undefined]} readOnly={false} required={[undefined]} tabIndex={[undefined]} aria-describedby={[undefined]} aria-invalid={[undefined]} aria-labelledby={[undefined]} aria-required={[undefined]} />
                                    </div>
                                  </div>
                                </InnerReference>
                              </Reference>
                            </Manager>
                          </r>
                        </r>
                      </td>
                    </tr>
                    <TableErrors errors=\\"\\">
                      <tr>
                        <td colSpan={2}>
                          <div className=\\"errorBox\\" />
                        </td>
                      </tr>
                    </TableErrors>
                    <tr>
                      <td className=\\"cellModal\\">
                        Room: 
                      </td>
                      <td>
                        <div style={{...}}>
                          <StateManager classNamePrefix=\\"list\\" className=\\"listRooms\\" styles={{...}} value={{...}} onChange={[Function: onChange]} name=\\"id_room\\" options={{...}} defaultInputValue=\\"\\" defaultMenuIsOpen={false} defaultValue={{...}}>
                            <Select classNamePrefix=\\"list\\" className=\\"listRooms\\" styles={{...}} value={{...}} onChange={[Function (anonymous)]} name=\\"id_room\\" options={{...}} inputValue=\\"\\" menuIsOpen={false} onInputChange={[Function (anonymous)]} onMenuClose={[Function (anonymous)]} onMenuOpen={[Function (anonymous)]} backspaceRemovesValue={true} blurInputOnSelect={true} captureMenuScroll={false} closeMenuOnSelect={true} closeMenuOnScroll={false} components={{...}} controlShouldRenderValue={true} escapeClearsValue={false} filterOption={[Function (anonymous)]} formatGroupLabel={[Function: formatGroupLabel]} getOptionLabel={[Function: getOptionLabel]} getOptionValue={[Function: getOptionValue]} isDisabled={false} isLoading={false} isMulti={false} isRtl={false} isSearchable={true} isOptionDisabled={[Function: isOptionDisabled]} loadingMessage={[Function: loadingMessage]} maxMenuHeight={300} minMenuHeight={140} menuPlacement=\\"bottom\\" menuPosition=\\"absolute\\" menuShouldBlockScroll={false} menuShouldScrollIntoView={true} noOptionsMessage={[Function: noOptionsMessage]} openMenuOnFocus={false} openMenuOnClick={true} pageSize={5} placeholder=\\"Select...\\" screenReaderStatus={[Function: screenReaderStatus]} tabIndex=\\"0\\" tabSelectsValue={true}>
                              <SelectContainer cx={[Function (anonymous)]} clearValue={[Function (anonymous)]} getStyles={[Function (anonymous)]} getValue={[Function (anonymous)]} hasValue={false} isMulti={false} isRtl={false} options={{...}} selectOption={[Function (anonymous)]} setValue={[Function (anonymous)]} selectProps={{...}} theme={{...}} className=\\"listRooms\\" innerProps={{...}} isDisabled={false} isFocused={false}>
                                <EmotionCssPropInternal css={{...}} className=\\"listRooms\\" id={[undefined]} onKeyDown={[Function (anonymous)]} __EMOTION_TYPE_PLEASE_DO_NOT_USE__=\\"div\\" __EMOTION_LABEL_PLEASE_DO_NOT_USE__=\\"SelectContainer\\">
                                  <div className=\\"listRooms css-jxd5du-container\\" id={[undefined]} onKeyDown={[Function (anonymous)]}>
                                    <Control cx={[Function (anonymous)]} clearValue={[Function (anonymous)]} getStyles={[Function (anonymous)]} getValue={[Function (anonymous)]} hasValue={false} isMulti={false} isRtl={false} options={{...}} selectOption={[Function (anonymous)]} setValue={[Function (anonymous)]} selectProps={{...}} theme={{...}} innerRef={[Function (anonymous)]} innerProps={{...}} isDisabled={false} isFocused={false} menuIsOpen={false}>
                                      <EmotionCssPropInternal css={{...}} className=\\"list__control\\" onMouseDown={[Function (anonymous)]} onTouchEnd={[Function (anonymous)]} __EMOTION_TYPE_PLEASE_DO_NOT_USE__=\\"div\\" __EMOTION_LABEL_PLEASE_DO_NOT_USE__=\\"Control\\">
                                        <div className=\\"list__control css-1ayw91x-control\\" onMouseDown={[Function (anonymous)]} onTouchEnd={[Function (anonymous)]}>
                                          <ValueContainer cx={[Function (anonymous)]} clearValue={[Function (anonymous)]} getStyles={[Function (anonymous)]} getValue={[Function (anonymous)]} hasValue={false} isMulti={false} isRtl={false} options={{...}} selectOption={[Function (anonymous)]} setValue={[Function (anonymous)]} selectProps={{...}} theme={{...}} isDisabled={false}>
                                            <EmotionCssPropInternal css={{...}} className=\\"list__value-container\\" __EMOTION_TYPE_PLEASE_DO_NOT_USE__=\\"div\\" __EMOTION_LABEL_PLEASE_DO_NOT_USE__=\\"ValueContainer\\">
                                              <div className=\\"list__value-container css-g1d714-ValueContainer\\">
                                                <Placeholder cx={[Function (anonymous)]} clearValue={[Function (anonymous)]} getStyles={[Function (anonymous)]} getValue={[Function (anonymous)]} hasValue={false} isMulti={false} isRtl={false} options={{...}} selectOption={[Function (anonymous)]} setValue={[Function (anonymous)]} selectProps={{...}} theme={{...}} isDisabled={false} isFocused={false}>
                                                  <EmotionCssPropInternal css={{...}} className=\\"list__placeholder\\" __EMOTION_TYPE_PLEASE_DO_NOT_USE__=\\"div\\" __EMOTION_LABEL_PLEASE_DO_NOT_USE__=\\"Placeholder\\">
                                                    <div className=\\"list__placeholder css-1wa3eu0-placeholder\\">
                                                      Select...
                                                    </div>
                                                  </EmotionCssPropInternal>
                                                </Placeholder>
                                                <Input autoCapitalize=\\"none\\" autoComplete=\\"off\\" autoCorrect=\\"off\\" cx={[Function (anonymous)]} getStyles={[Function (anonymous)]} id=\\"react-select-2-input\\" innerRef={[Function (anonymous)]} isDisabled={false} isHidden={false} onBlur={[Function (anonymous)]} onChange={[Function (anonymous)]} onFocus={[Function (anonymous)]} selectProps={{...}} spellCheck=\\"false\\" tabIndex=\\"0\\" form={[undefined]} theme={{...}} type=\\"text\\" value=\\"\\" aria-autocomplete=\\"list\\" aria-label={[undefined]} aria-labelledby={[undefined]}>
                                                  <EmotionCssPropInternal css={{...}} __EMOTION_TYPE_PLEASE_DO_NOT_USE__=\\"div\\" __EMOTION_LABEL_PLEASE_DO_NOT_USE__=\\"Input\\">
                                                    <div className=\\"css-b8ldur-Input\\">
                                                      <AutosizeInput className=\\"list__input\\" inputRef={[Function (anonymous)]} inputStyle={{...}} disabled={false} autoCapitalize=\\"none\\" autoComplete=\\"off\\" autoCorrect=\\"off\\" id=\\"react-select-2-input\\" onBlur={[Function (anonymous)]} onChange={[Function (anonymous)]} onFocus={[Function (anonymous)]} spellCheck=\\"false\\" tabIndex=\\"0\\" form={[undefined]} type=\\"text\\" value=\\"\\" aria-autocomplete=\\"list\\" aria-label={[undefined]} aria-labelledby={[undefined]} minWidth={1} injectStyles={true}>
                                                        <div className=\\"list__input\\" style={{...}}>
                                                          <input className={[undefined]} disabled={false} autoCapitalize=\\"none\\" autoComplete=\\"off\\" autoCorrect=\\"off\\" id=\\"react-select-2-input\\" onBlur={[Function (anonymous)]} onChange={[Function (anonymous)]} onFocus={[Function (anonymous)]} spellCheck=\\"false\\" tabIndex=\\"0\\" form={[undefined]} type=\\"text\\" value=\\"\\" aria-autocomplete=\\"list\\" aria-label={[undefined]} aria-labelledby={[undefined]} style={{...}} />
                                                          <div style={{...}} />
                                                        </div>
                                                      </AutosizeInput>
                                                    </div>
                                                  </EmotionCssPropInternal>
                                                </Input>
                                              </div>
                                            </EmotionCssPropInternal>
                                          </ValueContainer>
                                          <IndicatorsContainer cx={[Function (anonymous)]} clearValue={[Function (anonymous)]} getStyles={[Function (anonymous)]} getValue={[Function (anonymous)]} hasValue={false} isMulti={false} isRtl={false} options={{...}} selectOption={[Function (anonymous)]} setValue={[Function (anonymous)]} selectProps={{...}} theme={{...}} isDisabled={false}>
                                            <EmotionCssPropInternal css={{...}} className=\\"list__indicators\\" __EMOTION_TYPE_PLEASE_DO_NOT_USE__=\\"div\\" __EMOTION_LABEL_PLEASE_DO_NOT_USE__=\\"IndicatorsContainer\\">
                                              <div className=\\"list__indicators css-1hb7zxy-IndicatorsContainer\\">
                                                <IndicatorSeparator cx={[Function (anonymous)]} clearValue={[Function (anonymous)]} getStyles={[Function (anonymous)]} getValue={[Function (anonymous)]} hasValue={false} isMulti={false} isRtl={false} options={{...}} selectOption={[Function (anonymous)]} setValue={[Function (anonymous)]} selectProps={{...}} theme={{...}} isDisabled={false} isFocused={false}>
                                                  <EmotionCssPropInternal css={{...}} className=\\"list__indicator-separator\\" __EMOTION_TYPE_PLEASE_DO_NOT_USE__=\\"span\\" __EMOTION_LABEL_PLEASE_DO_NOT_USE__=\\"IndicatorSeparator\\">
                                                    <span className=\\"list__indicator-separator css-1okebmr-indicatorSeparator\\" />
                                                  </EmotionCssPropInternal>
                                                </IndicatorSeparator>
                                                <DropdownIndicator cx={[Function (anonymous)]} clearValue={[Function (anonymous)]} getStyles={[Function (anonymous)]} getValue={[Function (anonymous)]} hasValue={false} isMulti={false} isRtl={false} options={{...}} selectOption={[Function (anonymous)]} setValue={[Function (anonymous)]} selectProps={{...}} theme={{...}} innerProps={{...}} isDisabled={false} isFocused={false}>
                                                  <EmotionCssPropInternal onMouseDown={[Function (anonymous)]} onTouchEnd={[Function (anonymous)]} aria-hidden=\\"true\\" css={{...}} className=\\"list__indicator list__dropdown-indicator\\" __EMOTION_TYPE_PLEASE_DO_NOT_USE__=\\"div\\" __EMOTION_LABEL_PLEASE_DO_NOT_USE__=\\"DropdownIndicator\\">
                                                    <div onMouseDown={[Function (anonymous)]} onTouchEnd={[Function (anonymous)]} aria-hidden=\\"true\\" className=\\"list__indicator list__dropdown-indicator css-tlfecz-indicatorContainer\\">
                                                      <DownChevron>
                                                        <Svg size={20}>
                                                          <EmotionCssPropInternal height={20} width={20} viewBox=\\"0 0 20 20\\" aria-hidden=\\"true\\" focusable=\\"false\\" css={{...}} __EMOTION_TYPE_PLEASE_DO_NOT_USE__=\\"svg\\" __EMOTION_LABEL_PLEASE_DO_NOT_USE__=\\"Svg\\">
                                                            <svg height={20} width={20} viewBox=\\"0 0 20 20\\" aria-hidden=\\"true\\" focusable=\\"false\\" className=\\"css-tj5bde-Svg\\">
                                                              <path d=\\"M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z\\" />
                                                            </svg>
                                                          </EmotionCssPropInternal>
                                                        </Svg>
                                                      </DownChevron>
                                                    </div>
                                                  </EmotionCssPropInternal>
                                                </DropdownIndicator>
                                              </div>
                                            </EmotionCssPropInternal>
                                          </IndicatorsContainer>
                                        </div>
                                      </EmotionCssPropInternal>
                                    </Control>
                                    <input name=\\"id_room\\" type=\\"hidden\\" value=\\"\\" />
                                  </div>
                                </EmotionCssPropInternal>
                              </SelectContainer>
                            </Select>
                          </StateManager>
                        </div>
                      </td>
                    </tr>
                    <TableErrors errors=\\"\\">
                      <tr>
                        <td colSpan={2}>
                          <div className=\\"errorBox\\" />
                        </td>
                      </tr>
                    </TableErrors>
                    <tr>
                      <td className=\\"cellModal\\">
                        Cost: 
                      </td>
                      <td>
                        <span className=\\"cost\\">
                          0
                        </span>
                         $
                      </td>
                    </tr>
                  </tbody>
                </table>
                <Button onClick={[Function: onClick]} className=\\" deleteButton deleteButton\\" title=\\"Delete\\">
                  <button onClick={[Function: onClick]} className=\\" deleteButton deleteButton\\">
                    Delete
                  </button>
                </Button>
                <div className=\\"modalBoxButton\\">
                  <Button className=\\"actionButton\\" onClick={[Function: onClick]} title=\\"Save\\">
                    <button onClick={[Function: onClick]} className=\\"actionButton\\">
                      Save
                    </button>
                  </Button>
                  <Button onClick={[Function: onClick]} className=\\"closeButton\\" title=\\"Cancel\\">
                    <button onClick={[Function: onClick]} className=\\"closeButton\\">
                      Cancel
                    </button>
                  </Button>
                </div>
              </div>
            </div>
          </ModalPortal>
        </Portal>
      </Modal>
    </div>
  </BookingModal>
</Provider>"
`;
